### PR TITLE
Block

### PR DESCRIPTION
**Pull request by jimchen5209**
## Prerequisites
- [x] #288 Merged
- [x] Wait for tick event merged and rewrite lock #282 
### 2020/03/30
- [x] #310 Merged for refactoring
### 2020/03/31
- [x] ~~#301~~ #302 Merged for testing
### 2020/06/07
- [x] #316 Merged for image

## Changes
- [x] Block Item
- [x] Place Block
- [x] Register block for action
- [x] Block event
- [x] Remove block
- [x] Cancel block event when activated
- [x] Already Server Thread
### 2020/03/30
- [x] Overwrite grass (Rewrite with another event)
- [x] Change to command block
  - [X] ~~Item Decreasment broke~~ (Works fine to me)
  - [x] Overwrite command block color
- [x] ~~Creative break block removal~~ Cancel custom block break event
- [x] ~~Add `Marker` Tag for no hitbox (TDB)~~ (Cancelled)
  - [X] ~~EntitySelecter need to be changed. (Finding ways other than Box(AABB)~~ (Cancelled)
### 2020/05/26
- [x] Check if there is a block before placing (Rewrite with another event)
### 2020/08/10
- [x] Interact entity

### 2020/01/05
- [x] ~~Water is placeable directly onto a custom block with GUI (and triggers GUI)~~
- [x] ~~Sword is not equipped onto the armor stand (no block texture)~~
- [x] ~~Custom block cannot be placed directly onto another custom block without GUI~~
### 2020/03/30
- [x] ~~Item blocking placement~~(Rewrite with another event)
### 2020/05/26
- [x] ~~Entity check not working~~
### 2020/06/06
- [x] ~~Block placed when using the target block~~(Rewrite with another event)
### 2020/06/07
- [X] ~~Harvestable block event triggers twice when using bonemeal~~ (Rewrite with another event)
### 2020/08/12
- [x] ~~Block break by explosion currently not cancellable~~

### 2020/08/09
- Doesn't work if player is creative mode with level two op permission
### 2020/08/10
- 物品放副手，主手拿扳手拆方塊時副手的東西有機會被使用，然後揮錯手

### CODE REVIEW

**Comment by jimchen5209** [comment]
- [ ] Water is placeable
- [ ] Sword is not equipped
- [ ] Block is unplaceable clicking a non gui custom block

**Comment by jimchen5209** [comment]
- [ ] Overwrite grass
- [ ] Change to command block
- [ ] Item blocking placement
- [ ] Creative break block removal

**Comment by jimchen5209** [comment]
- [ ] #301 Merged for testing

**Comment by jimchen5209** [comment]
- [ ] Hand swings when fail to place block

**Comment by jimchen5209** [comment]
- [ ] Block placed when using the target block

**Comment by jimchen5209** [comment]
- [ ] Harvest triggered twice when using bonemeal

**Comment by jimchen5209** [comment]
Doesn't work with creative mode with op permission

File: src/main/kotlin/one/oktw/galaxy/block/type/BlockType.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package one.oktw.galaxy.block.type
+
+enum class BlockType(val customModelData: Int?, val languageKey: String, val hasGUI: Boolean = false) {
```
Comment (james58899) [inline src/main/kotlin/one/oktw/galaxy/block/type/BlockType.kt:21]: 為什麼 `customModelData` 是nullable?

Comment (jimchen5209) [reply src/main/kotlin/one/oktw/galaxy/block/type/BlockType.kt:21]: 圖片

Comment (jimchen5209) [reply src/main/kotlin/one/oktw/galaxy/block/type/BlockType.kt:21]: 星球終端跟大廳的那個主控台沒有材質（應該說是註冊到其他實體方塊的）

File: src/main/kotlin/one/oktw/galaxy/block/item/BlockItem.kt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+class BlockItem(val type: BlockType = DUMMY) : Item {
+    override val itemType = BLOCK
+
+    override val baseItem: net.minecraft.item.Item = COMMAND_BLOCK
+
+    override fun createItemStack(): ItemStack {
+        val item = CustomItemBuilder()
+            .setBaseItem(baseItem)
+            .setModel(type.customModelData!!)
```
Comment (james58899) [inline src/main/kotlin/one/oktw/galaxy/block/item/BlockItem.kt:39]: 應該要判斷是否有 `customModelData`

File: src/main/kotlin/one/oktw/galaxy/block/Block.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import one.oktw.galaxy.block.type.BlockType
+import one.oktw.galaxy.block.type.BlockType.DUMMY
+import one.oktw.galaxy.block.util.CustomBlockUtil
+
+class Block(val type: BlockType = DUMMY) {
+    val item = if (type.customModelData != 0 || type.name == "DUMMY") BlockItem(type) else null
+
+    fun register(world: ServerWorld, blockPos: BlockPos) = CustomBlockUtil.registerBlock(world, blockPos, type)
+
+    fun place(context: ItemUsageContext) =
```
Comment (james58899) [inline src/main/kotlin/one/oktw/galaxy/block/Block.kt:34]:
```suggestion
    fun place(context: ItemPlacementContext) =
```
直接用 `ItemPlacementContext` 比較合理

**Comment by jimchen5209** [comment]
# 已知問題：
- 物品放副手，主手拿扳手拆方塊時副手的東西有機會被使用，然後揮錯手

**Comment by jimchen5209** [comment]
- [ ] Interact entity

**Comment by jimchen5209** [comment]
- [ ] Block break by explosion currently not cancellable

File: src/main/java/one/oktw/galaxy/mixin/event/MixinBlockBreak_Explosion.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    @Redirect(method = "collectBlocksAndDamageEntities", at = @At(
+        value = "INVOKE",
+        target = "Ljava/util/List;addAll(Ljava/util/Collection;)Z",
+        ordinal = 0
+    ))
+    private boolean onDestroyByExplosion(List<BlockPos> list, Collection<? extends BlockPos> affectedPos) {
+        Main main = Main.Companion.getMain();
+        if (main == null) return list.addAll(affectedPos);
+        Set<BlockPos> set = Sets.newHashSet();
+        affectedPos.forEach(blockPos -> {
```
[the remark this answers is not in the document]
Comment (jimchen5209) [reply src/main/java/one/oktw/galaxy/mixin/event/MixinBlockBreak_Explosion.java:53]: event 跟其他破壞事件一樣都是一個一個傳
而且在 list.forEach 裡面改 list 會炸開

[the remark this answers is not in the document]
Comment (james58899) [reply src/main/java/one/oktw/galaxy/mixin/event/MixinBlockBreak_Explosion.java:53]: 那就讓它獨立一個event，這樣太沒效率了

[the remark this answers is not in the document]
Comment (jimchen5209) [reply src/main/java/one/oktw/galaxy/mixin/event/MixinBlockBreak_Explosion.java:53]: 傳進去再 forEach 會比較有效率嗎 🤔

[the remark this answers is not in the document]
Comment (james58899) [reply src/main/java/one/oktw/galaxy/mixin/event/MixinBlockBreak_Explosion.java:53]: 把東西複製兩次比原地刪除還要慢很多

**Review by james58899 — APPROVED** [review]
我想merge這個PR

**Pull request merged**